Modernize project

0xced opened the pull request:
1. Update target frameworks
* Drop .NET Standard below 2.0
* Update .NET Framework from 4.5 to 4.6.2
* Add .NET 6

2. Cleanup projects and code

3. Add global.json

davidebbo commented on src/ReflectionMagic/ReflectionMagic.csproj:
```diff
@@ -3,8 +3,7 @@
     <VersionPrefix>4.1.0</VersionPrefix>
-    <TargetFrameworks>net45;netstandard1.5;netstandard2.0</TargetFrameworks>
-    
+    <TargetFrameworks>net6.0;net462;netstandard2.0</TargetFrameworks>
```
On my default clean install, I don't have 4.6.2:
    
    MSB3644	The reference assemblies for .NETFramework,Version=v4.6.2 were not found. To resolve this, install the Developer Pack (SDK/Targeting Pack) for this framework version or retarget your application...

Is there a downside to going to 4.8? Maybe the downside is to accidentally use 4.8 things, hence making it unusable in 4.6.*?

Bear with me, my .net skills are very rusty!

0xced replied:
Following Microsoft practices, most NuGet packages with support for .NET Framework are currently targeting .NET Framework 4.6.2 (which will be [supported until Jan 12, 2027][1]). Since this project does not take advantage of APIs introduced in .NET Framework 4.8, I'd stick with supporting .NET Framework 4.6.2.

The error about reference assemblies not found is weird. I have added a `global.json` file that will force using at least the .NET 6.0.400 SDK, where [references assemblies are automatically downloaded][2]. This automatic download behaviour exists since the .NET 5 SDK.

What does `dotnet --info` say when run inside the ReflectionMagic directory?

[1]: https://learn.microsoft.com/en-us/lifecycle/products/microsoft-net-framework
[2]: https://github.com/dotnet/sdk/pull/10981

davidebbo replied:
Note that the error above is when building from VS. If I run `dotnet build` from cmd line, I get:

```
global.json file: D:\repos\ReflectionMagic\global.json

Installed SDKs:
7.0.201 [C:\Program Files\dotnet\sdk]

Install the [6.0.400] .NET SDK or update [D:\repos\ReflectionMagic\global.json] to match an installed SDK.
```

jvandertil replied:
That's due to the global.json specifying that it requires that specific SDK version. I don't think we really need it, but it does make things more reproducible on the build server side.

davidebbo replied:
I installed dotnet core 6.0, and that allowed building using `dotnet build`. It still fails the same from VS. I suspect I could fix it with the dev pack.

I'm a bit confused about the build system from VS vs command line. Are they two totally different things, or did they unify them? A few years ago (last time I used .net), I think a project had to be either .NET Core or .NET full framework, but not both.

davidebbo replied:
Installing the dev pack fixes the VS build, so everything works now.

@jvandertil other than the required SDK version (which we can discuss further), are you good with this change?

jvandertil replied:
Yes, LGTM.

0xced commented on src/ReflectionMagic/Properties/AssemblyInfo.cs:
```diff
@@ -1,19 +0,0 @@
-[assembly: ComVisible(false)]
-
-// The following GUID is for the ID of the typelib if this project is exposed to COM
-[assembly: Guid("3c576b3b-09a9-49a6-a698-5b693c0005cc")]
```
Yes, exactly. The [MSBuild reference for .NET SDK projects][1] has a table with a mapping between MSBuild properties and the corresponding assembly attributes.

[1]: https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#assembly-attribute-properties

davidebbo replied:
Awesome.

0xced commented:
I have made several other modernizations:
* Bumped version to 5.0.0 (because dropping target frameworks is a breaking change)
* Improved NuGet packaging (add the [README file in the pacakge](https://devblogs.microsoft.com/nuget/add-a-readme-to-your-nuget-package/) and use non deprecated MSBuild properties)
* Removed obsolete methods. It's been 4 years and nobody complained (the obsolete warning explicitly asked to open an issue on GitHub)
* Added [SourceLink](https://devblogs.microsoft.com/dotnet/producing-packages-with-source-link/) support

I can also help if you want to setup GitHub actions to build + test + deploy a new version of the NuGet package.

davidebbo commented:
@0xced Doing the GitHub action would be great, and good learning for me. BTW, I emailed you yesterday, asking if you wanted to become contributor, given the limited active support that I can provide.

jvandertil commented:
I already have an Azure pipelines for this, but I should modernize it a bit. Maybe a GitHub pipeline makes more sense, but we'll see. I'll give this a run through locally and then push out a new version to NuGet later today :)
